x/shield/keeper/proposal.go

```diff
@@ -17,3 +21,339 @@ func (k Keeper) ClaimUnlock(ctx sdk.Context, proposalID, poolID uint64, loss sdk
+		var ubdAmountDec sdk.Dec
+		if totalDelAmountDec.GT(payoutDec) {
+			// FIXME: Corner case: not enough amount in the last delegation.
+			if i == len(delegations)-1 {
+				ubdAmountDec = remainingDec
+			} else {
+				ubdAmountDec = payoutDec.Mul(delAmountDec).Quo(totalDelAmountDec)
+				remainingDec = remainingDec.Sub(ubdAmountDec)
+			}
+		} else {
+			ubdAmountDec = delAmountDec
+		}
+		ubdShares, err := val.SharesFromTokens(ubdAmountDec.TruncateInt())
+		if err != nil {
+			panic(err)
+		}
+		k.UndelegateShares(ctx, delegations[i].DelegatorAddress, delegations[i].ValidatorAddress, ubdShares)
+	}
+}
+
+// PayFromUnbonding reduce provider's unbonding delegation and transfer tokens to the shield module account.
+func (k Keeper) PayFromUnbonding(ctx sdk.Context, ubd staking.UnbondingDelegation, payout sdk.Int) {
+	delAddr := ubd.DelegatorAddress
+	valAddr := ubd.ValidatorAddress
+	unbonding, found := k.sk.GetUnbondingDelegation(ctx, delAddr, valAddr)
+	if !found {
+		panic("unbonding delegation is not found")
+	}
+
+	// Update unbonding delegations between the delegator and the validator.
+	for i := range unbonding.Entries {
+		if unbonding.Entries[i].Balance.Equal(ubd.Entries[0].Balance) && unbonding.Entries[i].CompletionTime.Equal(ubd.Entries[0].CompletionTime) {
+			unbonding.Entries[i].Balance = unbonding.Entries[i].Balance.Sub(payout)
+			k.sk.SetUnbondingDelegation(ctx, unbonding)
+			break
+		}
+	}
+
+	// FIXME: Update the unbonding queue only if entry is removed.
+
+	// Transfer tokens from the staking module to the shield module.
+	payoutCoins := sdk.NewCoins(sdk.NewCoin(k.sk.BondDenom(ctx), payout))
+	if err := k.supplyKeeper.SendCoinsFromModuleToModule(ctx, staking.NotBondedPoolName, types.ModuleName, payoutCoins); err != nil {
+		panic(err)
+	}
+}
+
+// UndelegateShares undelegates delegations of a delegator to a validator by shares.
+func (k Keeper) UndelegateShares(ctx sdk.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress, shares sdk.Dec) {
+	delegation, found := k.sk.GetDelegation(ctx, delAddr, valAddr)
+	if !found {
+		panic("delegation is not found")
+	}
+	k.sk.BeforeDelegationSharesModified(ctx, delAddr, valAddr)
+
+	// Undelegate coins from the staking module account to the shield module account.
+	validator, found := k.sk.GetValidator(ctx, valAddr)
+	if !found {
+		panic("validator was not found")
+	}
+	ubdCoins := sdk.NewCoins(sdk.NewCoin(k.sk.BondDenom(ctx), validator.TokensFromShares(shares).TruncateInt()))
+	if err := k.supplyKeeper.SendCoinsFromModuleToModule(ctx, staking.BondedPoolName, types.ModuleName, ubdCoins); err != nil {
+		panic(err)
+	}
+
+	// Update delegation records.
+	delegation.Shares = delegation.Shares.Sub(shares)
+	k.sk.SetDelegation(ctx, delegation)
+	k.sk.AfterDelegationModified(ctx, delegation.DelegatorAddress, delegation.ValidatorAddress)
+
+	// Update the validator.
+	k.sk.RemoveValidatorTokensAndShares(ctx, validator, shares)
```

Comment:
Should we follow `unbond()` logic here technically? `unbond()`, removes a delegation and a validator if `delegation.Shares.IsZero()` and `validator.DelegatorShares.IsZero()` are true, respectively.

Reply:
`unbond()` also triggers jailing if self delegation falls below minimum.

Reply:
Updated.